src/arizona_template_compiler.erl

```diff
@@ -0,0 +1,653 @@
+                    | invalid_block_name
+                    | invalid_macro_value,
+         ErrMod :: module(),
+         ErrFun :: atom(),
+         ErrLoc :: arizona_template_scanner:location().
+compile(Mod, Fun, Macros) ->
+    try
+        State = #state{
+            module = Mod,
+            function = Fun,
+            macros = Macros,
+            attributes = [],
+            index = 0,
+            path = []
+        },
+        Tree = expand_block(Mod, Fun, _Attrs = [], State),
+        {ok, block(Tree, _NormAssigns = #{}, State)}
+    catch
+        % That's not the location in the module but the location
+        % in the template starting from {1, 1}.
+        throw:{ErrReason, ErrLoc, ErrState} ->
+            ErrMod = ErrState#state.module,
+            ErrFun = ErrState#state.function,
+            {error, {ErrReason, {ErrMod, ErrFun, ErrLoc}}}
+    end.
+
+%% --------------------------------------------------------------------
+%% Private
+%% --------------------------------------------------------------------
+
+compile_elems([{text, _Loc, Txt} | T], State) ->
+    compile_text(Txt, T, State);
+compile_elems([{expr, Loc, Expr} | T], State) ->
+    compile_expr(Expr, Loc, T, State);
+compile_elems([{block, Loc, Block} | T], State) ->
+    compile_block(Block, Loc, T, State);
+compile_elems([{tag, Loc, #{is_stateful := true}} | _T], State) ->
+    throw({unexpected_stateful_tag, Loc, State});
+compile_elems([{tag, Loc, Tag} | T], State) ->
+    compile_tag(Tag, Loc, T, State);
+compile_elems([], _State) ->
+    [].
+
+compile_text(Txt, T, State) ->
+    [{text, Txt} | compile_elems(T, State)].
+
+compile_expr(ExprStr, Loc, T, State) ->
+    case expand_expr(ExprStr, Loc, true, State) of
+        {text, Txt} ->
+            [{text, Txt} | compile_elems(T, State)];
+        {expr, Expr} ->
+            [{expr, Expr} | compile_elems(T, incr_index(1, State))]
+    end.
+
+expand_expr(Expr, Loc, Eval, State) ->
+    ExprTree = merl:quote(Expr),
+    AttrsEnv = lists:map(fun
+        ({K, {expr, _KLoc, KExpr}}) ->
+            {binary_to_existing_atom(K, utf8), merl:quote(KExpr)};
+        ({K, {text, _KLoc, KTxt}}) ->
+            {binary_to_existing_atom(K, utf8),
+              merl:quote(iolist_to_binary(["<<\"", KTxt, "\"/utf8>>"]))}
+    end, State#state.attributes),
+    AttrsTree = merl:tsubst(ExprTree, AttrsEnv),
+    Macros = State#state.macros,
+    MacrosEnv = [{K, merl:term(V)} || K := V <- Macros],
+    MacrosTree = merl:tsubst(AttrsTree, MacrosEnv),
+    AllVars = merl:template_vars(merl:template(MacrosTree)),
+    case AllVars -- maps:keys(Macros) of
+        [] when Eval ->
+            try
+                Form = erl_syntax:revert(merl:tree(MacrosTree)),
+                {value, Value, _} = erl_eval:exprs([Form], []),
+                {text, arizona_html:to_safe(Value)}
+            catch
+                _:_ ->
+                    throw({invalid_macro_value, Loc, State})
+            end;
```

Comment:
Could this be made into a function so we isolate the try-catch even further?

Reply:
[e8e5ccd](https://github.com/arizona-framework/arizona/pull/164/commits/e8e5ccdfc925ebcb18cd4c32b47be047f9034a76)